strawberryfields/backends/tfbackend/ops.py

```diff
@@ -931,24 +931,11 @@ def replace_modes(replacement, modes, system, system_is_pure, batched=False):
         # append mode and insert state (There is also insert_state(), but it seemed unnecessarily complicated to try to generalize this function, which does a lot of manual list comprehension, to the multi mode case than to write the two liner below)
         # todo: insert_state() could be rewritten to take full advantage of the high level functions of tf. Before doing that different implementatinos should be benchmarked to compare speed and memory requirements, as in practice these methods will be perfomance critical.
         if not batched:
-            # todo: remove the hack in the line below and enable the line with axes=0 instead, if ever we change the dependency of SF to tensorflow>=1.6
-            # revised_modes = tf.tensordot(reduced_state, replacement, axes=0)
-            revised_modes = tf.tensordot(
-                tf.expand_dims(reduced_state, 0), tf.expand_dims(replacement, 0), axes=[[0], [0]]
-            )
+            revised_modes = tf.tensordot(reduced_state, replacement, axes=0)
```

Comment:
Nice! This should also help with speed I imagine?

I'm just tagging @ziofil for any thoughts or comments.

Reply:
I made a few experiments and I found that broadcasting is about 40x faster than the method above. it can be done programmatically by writing the indices like so:
```python
num_modes = len(reduced_state.shape) - 1
idx1 = (slice(None,None),)+(Ellipsis,) + (None,)* num_modes
idx2 = (slice(None,None),)+(None,)* num_modes +(Ellipsis,)
revised_modes = reduced_state[idx1] * replacement[idx2]
```

Reply:
one could even bypass the `if batched / else` block with 
```python
idx1 = (slice(None,None),)*batched + (Ellipsis,) + (None,)*(replacement.ndim-int(batched))
idx2 = (slice(None,None),)*batched + (None,)*(reduced_state.ndim-int(batched)) + (Ellipsis,)
revised_modes = reduced_state[idx1] * replacement[idx2]
```

Reply:
Oh, that's awesome. Is this generally the case that broadcasting is faster than tensordot? Or just for this particular application?

Reply:
Made and pushed these changes. Certainly a speedup in general, though specific cases are about equivalent (e.g. homodyne).